File: src/test/java/cz/vhromada/catalog/service/MusicServiceTest.java
package cz.vhromada.catalog.service;

import static org.assertj.core.api.Assertions.assertThatThrownBy;

import java.util.Collections;

import cz.vhromada.catalog.domain.Music;
import cz.vhromada.catalog.domain.Song;
import cz.vhromada.catalog.repository.MusicRepository;
import cz.vhromada.catalog.utils.MusicUtils;
import cz.vhromada.catalog.utils.SongUtils;
import cz.vhromada.common.repository.MovableRepository;
import cz.vhromada.common.service.MovableService;
import cz.vhromada.common.test.service.MovableServiceTest;

import org.junit.jupiter.api.Test;
import org.mockito.Mock;
import org.springframework.cache.Cache;

/**
 * A class represents test for class {@link MusicService}.
 *
 * @author Vladimir Hromada
 */
class MusicServiceTest extends MovableServiceTest<Music> {

    /**
     * Instance of {@link MusicRepository}
     */
    @Mock
    private MusicRepository repository;

    /**
     * Test method for {@link MusicService#MusicService(MusicRepository, Cache)} with null repository for music.
     */
    @Test
    void constructor_NullMusicRepository() {
        assertThatThrownBy(() -> new MusicService(null, getCache())).isInstanceOf(IllegalArgumentException.class);
    }

    /**
     * Test method for {@link MusicService#MusicService(MusicRepository, Cache)} with null cache.
     */
    @Test
    void constructor_NullCache() {
        assertThatThrownBy(() -> new MusicService(repository, null)).isInstanceOf(IllegalArgumentException.class);
    }

    @Override
    protected MovableRepository<Music> getRepository() {
        return repository;
    }

    @Override
    protected MovableService<Music> getService() {
        return new MusicService(repository, getCache());
    }

    @Override
    protected String getCacheKey() {
        return "music";
    }

    @Override
    protected Music getItem1() {
        return setSongs(MusicUtils.newMusicDomain(1));
    }

    @Override
    protected Music getItem2() {
        return setSongs(MusicUtils.newMusicDomain(2));
    }

    @Override
    protected Music getAddItem() {
        return MusicUtils.newMusicDomain(null);
    }

    @Override
    protected Music getCopyItem() {
        final Music music = MusicUtils.newMusicDomain(null);
        music.setPosition(0);
        setSongs(music);

        return music;
    }

    @Override
    protected Class<Music> getItemClass() {
        return Music.class;
    }

    @Override
    protected void assertDataDeepEquals(final Music expected, final Music actual) {
        MusicUtils.assertMusicDeepEquals(expected, actual);
    }

    /**
     * Sets songs to music.
     *
     * @param music music
     * @return music with songs
     */
    private static Music setSongs(final Music music) {
        final Integer id = music.getId();
        final Song song = SongUtils.newSongDomain(id);
        if (id == null) {
            song.setPosition(0);
        }
        music.setSongs(Collections.singletonList(song));

        return music;
    }

}

File: src/test/kotlin/cz/vhromada/catalog/service/MusicServiceTest.kt
package cz.vhromada.catalog.service

import com.nhaarman.mockitokotlin2.KArgumentCaptor
import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.argumentCaptor
import com.nhaarman.mockitokotlin2.verify
import com.nhaarman.mockitokotlin2.verifyNoMoreInteractions
import com.nhaarman.mockitokotlin2.whenever
import cz.vhromada.catalog.domain.Music
import cz.vhromada.catalog.repository.MusicRepository
import cz.vhromada.catalog.utils.MusicUtils
import cz.vhromada.catalog.utils.SongUtils
import cz.vhromada.common.provider.AccountProvider
import cz.vhromada.common.provider.TimeProvider
import cz.vhromada.common.service.MovableService
import cz.vhromada.common.test.service.MovableServiceTest
import cz.vhromada.common.test.utils.TestConstants
import org.mockito.Mock
import org.springframework.data.jpa.repository.JpaRepository

/**
 * A class represents test for class [MusicService].
 *
 * @author Vladimir Hromada
 */
class MusicServiceTest : MovableServiceTest<Music>() {

    /**
     * Instance of [MusicRepository]
     */
    @Mock
    private lateinit var repository: MusicRepository

    /**
     * Instance of [AccountProvider]
     */
    @Mock
    private lateinit var accountProvider: AccountProvider

    /**
     * Instance of [TimeProvider]
     */
    @Mock
    private lateinit var timeProvider: TimeProvider

    override fun getRepository(): JpaRepository<Music, Int> {
        return repository
    }

    override fun getAccountProvider(): AccountProvider {
        return accountProvider
    }

    override fun getTimeProvider(): TimeProvider {
        return timeProvider
    }

    override fun getService(): MovableService<Music> {
        return MusicService(repository, accountProvider, timeProvider, cache)
    }

    override fun getCacheKey(): String {
        return "music${TestConstants.ACCOUNT.id}"
    }

    override fun getItem1(): Music {
        return setSongs(MusicUtils.newMusicDomain(1))
    }

    override fun getItem2(): Music {
        return setSongs(MusicUtils.newMusicDomain(2))
    }

    override fun getAddItem(): Music {
        return MusicUtils.newMusicDomain(null)
    }

    override fun getCopyItem(): Music {
        val music = MusicUtils.newMusicDomain(null)
                .copy(position = 0)

        return setSongs(music)
    }

    override fun initAllDataMock(data: List<Music>) {
        whenever(repository.findByAuditCreatedUser(any())).thenReturn(data)
    }

    override fun verifyAllDataMock() {
        verify(repository).findByAuditCreatedUser(TestConstants.ACCOUNT_ID)
        verifyNoMoreInteractions(repository)
    }

    override fun anyItem(): Music {
        return any()
    }

    override fun argumentCaptorItem(): KArgumentCaptor<Music> {
        return argumentCaptor()
    }

    override fun assertDataDeepEquals(expected: Music, actual: Music) {
        MusicUtils.assertMusicDeepEquals(expected, actual)
    }

    /**
     * Sets songs to music.
     *
     * @param music music
     * @return music with songs
     */
    private fun setSongs(music: Music): Music {
        val id = music.id
        val song = SongUtils.newSongDomain(id)
        if (id == null) {
            song.position = 0
        }

        return music.copy(songs = listOf(song))
    }

}
